File: backend/services/fleet_api/lib/models/driver.py
from pydantic import BaseModel, Field, EmailStr
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class DriverStatus(str, Enum):
    """운전자 상태"""

    ACTIVE = "active"  # 활성
    INACTIVE = "inactive"  # 비활성
    SUSPENDED = "suspended"  # 정지
    ON_LEAVE = "on_leave"  # 휴가
    TRAINING = "training"  # 교육중
    TERMINATED = "terminated"  # 퇴사


class LicenseType(str, Enum):
    """운전면허 종류"""

    CLASS1 = "1종보통"  # 1종보통 (테스트에서 사용)
    CLASS2 = "2종보통"  # 2종보통 (테스트에서 사용)
    TYPE_1_REGULAR = "1종보통"  # 1종보통
    TYPE_1_LARGE = "1종대형"  # 1종대형
    TYPE_1_SPECIAL = "1종특수"  # 1종특수
    TYPE_2_REGULAR = "2종보통"  # 2종보통
    TYPE_2_SMALL = "2종소형"  # 2종소형
    TYPE_2_MOTORCYCLE = "2종원동기"  # 2종원동기
    CONSTRUCTION_MACHINERY = "건설기계"  # 건설기계


class EmergencyContact(BaseModel):
    """비상 연락처 모델"""

    name: str = Field(..., description="비상 연락처 이름")
    phone: str = Field(..., description="비상 연락처 전화번호")
    relationship: str = Field(..., description="비상 연락처 관계")


class DriverBase(BaseModel):
    """운전자 기본 모델"""

    name: str = Field(..., description="운전자 이름")
    email: EmailStr = Field(..., description="운전자 이메일")
    phone: str = Field(..., description="운전자 전화번호")
    license_number: str = Field(..., description="운전 면허 번호")
    license_type: Optional[str] = Field(None, description="운전 면허 종류")
    license_expiry: datetime = Field(..., description="운전 면허 만료일")

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "name": "홍길동",
                    "email": "driver@example.com",
                    "phone": "010-1234-5678",
                    "license_number": "123-45-678900",
                    "license_type": "1종보통",
                    "license_expiry": "2025-12-31T00:00:00Z",
                }
            ]
        }
    }


class DriverCreate(DriverBase):
    """운전자 생성 모델"""

    organization_id: str = Field(..., description="소속 조직 ID")
    is_active: bool = Field(True, description="활성 상태 여부")
    restrictions: Optional[List[str]] = Field(None, description="운전 제한사항")
    notes: Optional[str] = Field(None, description="비고")
    emergency_contact: Optional[EmergencyContact] = Field(
        None, description="비상 연락처"
    )
    address: Optional[str] = Field(None, description="주소")
    birth_date: Optional[datetime] = Field(None, description="생년월일")
    hire_date: Optional[datetime] = Field(None, description="입사일")
    department: Optional[str] = Field(None, description="부서")
    position: Optional[str] = Field(None, description="직책")
    profile_image_url: Optional[str] = Field(None, description="프로필 이미지 URL")


class DriverUpdate(BaseModel):
    """운전자 업데이트 모델"""

    name: Optional[str] = None
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    license_number: Optional[str] = None
    license_type: Optional[str] = None
    license_expiry: Optional[datetime] = None
    is_active: Optional[bool] = None
    restrictions: Optional[List[str]] = None
    notes: Optional[str] = None
    emergency_contact: Optional[EmergencyContact] = None
    address: Optional[str] = None
    birth_date: Optional[datetime] = None
    hire_date: Optional[datetime] = None
    department: Optional[str] = None
    position: Optional[str] = None
    profile_image_url: Optional[str] = None
    organization_id: Optional[str] = None


class DriverResponse(DriverBase):
    """운전자 응답 모델"""

    id: str
    organization_id: str
    is_active: bool
    restrictions: Optional[List[str]] = None
    notes: Optional[str] = None
    assigned_vehicles: Optional[List[str]] = None
    emergency_contact: Optional[EmergencyContact] = None
    address: Optional[str] = None
    birth_date: Optional[datetime] = None
    hire_date: Optional[datetime] = None
    department: Optional[str] = None
    position: Optional[str] = None
    profile_image_url: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}  # Pydantic v2 형식


class DriverFilters(BaseModel):
    """운전자 필터링 옵션"""

    name: Optional[str] = None
    email: Optional[str] = None
    license_number: Optional[str] = None
    is_active: Optional[bool] = None
    organization_id: Optional[str] = None
    license_expiry_before: Optional[datetime] = None
    license_expiry_after: Optional[datetime] = None
    page: int = Field(1, ge=1, description="페이지 번호")
    page_size: int = Field(10, ge=1, le=100, description="페이지 크기")


class DriverListResponse(BaseModel):
    """운전자 목록 응답"""

    items: List[DriverResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: packages/types/schema/driver.ts
/**
 * 드라이버(탁송 기사) 타입 정의
 */

/**
 * 드라이버 기본 정보
 */
export interface Driver {
  /** 드라이버 고유 ID */
  id: string;

  /** 드라이버 이름 */
  name: string;

  /** 이메일 주소 */
  email: string;

  /** 전화번호 */
  phone: string;

  /** 운전면허 번호 */
  licenseNumber: string;

  /** 운전면허 유형 */
  licenseType?: string;

  /** 운전면허 만료일 */
  licenseExpiry: string;

  /** 활성 상태 */
  isActive: boolean;

  /** 제한 사항 */
  restrictions?: string[];

  /** 메모 */
  notes?: string;

  /** 할당된 차량 목록 */
  assignedVehicles?: string[];

  /** 비상 연락처 */
  emergencyContact?: {
    name: string;
    phone: string;
    relationship: string;
  };

  /** 주소 */
  address?: string;

  /** 생년월일 */
  birthDate?: string;

  /** 고용일 */
  hireDate?: string;

  /** 부서 */
  department?: string;

  /** 직위 */
  position?: string;

  /** 프로필 이미지 URL */
  profileImageUrl?: string;

  /** 계정 생성일 */
  createdAt?: string;

  /** 마지막 업데이트 날짜 */
  updatedAt?: string;
}

/**
 * 드라이버 생성 입력
 */
export interface CreateDriverInput {
  /** 드라이버 이름 */
  name: string;

  /** 이메일 주소 */
  email: string;

  /** 전화번호 */
  phone: string;

  /** 운전면허 번호 */
  licenseNumber: string;

  /** 운전면허 만료일 */
  licenseExpiry: string;

  /** 운전면허 유형 */
  licenseType?: string;

  /** 제한 사항 */
  restrictions?: string[];

  /** 메모 */
  notes?: string;

  /** 활성 상태 */
  isActive?: boolean;

  /** 부서 */
  department?: string;

  /** 직위 */
  position?: string;

  /** 비상 연락처 */
  emergencyContact?: {
    name: string;
    phone: string;
    relationship: string;
  };
}

/**
 * 드라이버 업데이트 입력
 */
export interface UpdateDriverInput {
  /** 드라이버 이름 */
  name?: string;

  /** 이메일 주소 */
  email?: string;

  /** 전화번호 */
  phone?: string;

  /** 운전면허 번호 */
  licenseNumber?: string;

  /** 운전면허 만료일 */
  licenseExpiry?: string;

  /** 운전면허 유형 */
  licenseType?: string;

  /** 제한 사항 */
  restrictions?: string[];

  /** 메모 */
  notes?: string;

  /** 활성 상태 */
  isActive?: boolean;

  /** 부서 */
  department?: string;

  /** 직위 */
  position?: string;

  /** 비상 연락처 */
  emergencyContact?: {
    name: string;
    phone: string;
    relationship: string;
  };
}

/**
 * 드라이버 필터
 */
export interface DriverFilters {
  /** 활성 상태 */
  isActive?: boolean;

  /** 운전면허 유형 */
  licenseType?: string;

  /** 부서 */
  department?: string;

  /** 검색어 */
  search?: string;

  /** 페이지 번호 */
  page?: number;

  /** 페이지 크기 */
  pageSize?: number;
}

/**
 * 드라이버 위치 정보
 */
export interface DriverLocation {
  /** 위도 */
  latitude: number;

  /** 경도 */
  longitude: number;

  /** 위치 정확도 (미터) */
  accuracy?: number;

  /** 타임스탬프 */
  timestamp: number;

  /** 고도 (미터) */
  altitude?: number;

  /** 고도 정확도 (미터) */
  altitudeAccuracy?: number;

  /** 진행 방향 (0-360도) */
  heading?: number;

  /** 속도 (미터/초) */
  speed?: number;
}

/**
 * 드라이버 상태 열거형
 */
export enum DriverStatus {
  /** 오프라인 (앱 종료 또는 로그아웃) */
  OFFLINE = 'OFFLINE',

  /** 온라인 (업무 가능) */
  AVAILABLE = 'AVAILABLE',

  /** 배송 중 */
  DELIVERING = 'DELIVERING',

  /** 휴식 중 */
  BREAK = 'BREAK',

  /** 일시 정지 (계정 문제 등) */
  SUSPENDED = 'SUSPENDED',
}

/**
 * 드라이버 배송 통계
 */
export interface DriverStats {
  /** 드라이버 ID */
  driverId: string;

  /** 총 배송 수 */
  totalDeliveries: number;

  /** 오늘 완료한 배송 수 */
  todayDeliveries: number;

  /** 총 주행 거리 (킬로미터) */
  totalDistance: number;

  /** 평균 배송 시간 (분) */
  averageDeliveryTime: number;

  /** 평균 고객 평점 (5점 만점) */
  averageRating: number;

  /** 통계 마지막 업데이트 시간 */
  lastUpdated: string;
}

/**
 * 드라이버 배송 기록
 */
export interface DeliveryHistory {
  /** 배송 ID */
  id: string;

  /** 드라이버 ID */
  driverId: string;

  /** 출발지 */
  origin: string;

  /** 목적지 */
  destination: string;

  /** 배송 시작 시간 */
  startTime: string;

  /** 배송 완료 시간 */
  endTime: string;

  /** 배송 상태 */
  status: DeliveryStatus;

  /** 고객 평점 (5점 만점) */
  customerRating?: number;

  /** 고객 피드백 */
  customerFeedback?: string;

  /** 주행 거리 (킬로미터) */
  distance: number;
}

/**
 * 배송 상태 열거형
 */
export enum DeliveryStatus {
  /** 할당됨 */
  ASSIGNED = 'ASSIGNED',

  /** 픽업 중 */
  PICKING_UP = 'PICKING_UP',

  /** 배송 중 */
  IN_TRANSIT = 'IN_TRANSIT',

  /** 배송 완료 */
  DELIVERED = 'DELIVERED',

  /** 배송 실패 */
  FAILED = 'FAILED',

  /** 취소됨 */
  CANCELED = 'CANCELED',
}
